Conflict-resolution-perf-improvements.md

```diff
@@ -0,0 +1,112 @@
+and the root is a hash of all leaf hash values. This can be easily implemented by simply iterating over the segment container
+in order to calculate the hash of individual entries and the combined hash.
+
+TRACKED BY: [ISPN-8412] https://issues.redhat.com/browse/ISPN-8412
+
+### Hash to use
+MurmurHash3?
+
+### Calculating the entry hash
+Simply hash the `.hashcode()` returned by the `InternalCacheEntry` implementation.
+
+QUESTION: Should we include Metadata in the hash so that versions are taken into account? Are two entries with the same
+key/value but different metadata equal? Currently CR just calls `.equals` on the `CacheEntry` instances and
+`AbstractInternalCacheEntry`'s provided `equals` method only compares the key/values. Is this a bug?
+
+### Calculating the tree
+The computation of non-leaf nodes should only occur at the start of CR, as the additional iteration of entries would
+adversely affect cache write operation performance. The cost of calculating invidual leaf node hashes should be minimal,
+depending on the hash used, so this could be computed actively or lazily.
+
+### Increasing Merkle Tree Depth
+A segment root with all entries being it's children is not very efficient when a segment contains a large number of values,
+as when an inconsistency is discover, it's still necessary to send all entries within the segment over the wire. Increasing
+the depth of the tree means that it's possible to perform more fine-grained consistency checks, with a larger tree depth
+and smaller amount of leaf nodes resulting in less entries having to be sent over the wire when inconsistencies do occur.
+
+#### Constant lookup/removal tree - Depth 3
+This is a simple implemenetation which adds an additional layer of depth to the tree in order to increase CR granularity:
+
+* A segment is further divided into `X` nodes, which are the root's childen
+  - These nodes represent a range of a cache's `key` hashcodes, e.g. 0..5, 6..10 etc
+  - `(hash(key) & Integer.MAX_VALUE) \ X`
+  - Where `X` could be configurable, but should probably just be an internal constant
+
+* Each node contains:
+    - A Map to store leaf nodes, which enables amortised constant lookup and to accomodate `.hashcode()` conflicts on keys
+    - A node hash field containing a hash of all the leaf node's hashes combined
+
+* The tree, excluding leafs, can be represented as a int and int[X] (Root+X hashes) in a RPC
+
+* During CR if two tree's root values don't match, it's then possible to compare all `X` hashes and only request the indexes
+of the array which have conflicting hashes. At which point the participating nodes only return the InternalCacheEntries
+associated with those nodes.
+
+* Once the hash tree has been created for CR, it's possible to cache the hashes until a subsequent write operation. Upon
+a write operation it's necessary to invalidate the root hash and the hash of the child node containing the entry. During
```

Comment:
> Who needs the cached values?

I was thinking for subsequent CR requests, with also an eye on potential other use cases of the Merkle tree such as StateTransfer. 


> Is it worth caching the children hashes if recomputing one child's value requires scanning all the keys in the segment anyway?

If the tree has a depth of 3, we don't need to scan all of the keys, just those that co-exist in nodes at depth 1 in the tree that contain modified entries. To compute the root hash (depth 0), we need to combine the hashes of all nodes at depth 1.

So given the tree: `0(1 (3, 4, 5), 2(6, 7))`, if an entry associated with node 1, say 3, is updated then it's only necessary for the hash of 1 and 0 to be recalculated. When computing the root (node 0), we just use the existing hash of node 2.

> Should the entries modified during conflict resolution be considered for the hashing at all?

No, I think we need to compute the tree at the start of CR and treat all operations that occur during CR as the latest value, which we currently do by utlising the `PUT_FOR_STATE_TRANSFER` flag.

Reply:
> So given the tree: 0(1 (3, 4, 5), 2(6, 7)), if an entry associated with node 1, say 3, is updated then it's only necessary for the hash of 1 and 0 to be recalculated. When computing the root (node 0), we just use the existing hash of node 2.

The problem is that you don't know which keys belong to which node without computing their hashes, so in order to recalculate the hash of 1 and 0, you need to compute the hashes of all the keys in the segment.

Reply:
An entries location in the tree is deterministic and does not depend on any other entries, to find an entries location in the tree you only need to compute the hash of that entries key . So it's possible to just invalidate the affected nodes in the tree when the write operations are received.

CR shouldn't be a common occurence anyway, so caching/invalidating the hashes is probably overkill if this is the only application of the tree. Invalidating the hash cache is a constant operation, but obviously there is still a perf impact that's probably not worth the payoff.

Reply:
The invalidation part is easy, I never disputed that. But I don't understand how you could recompute the invalidated nodes' hashes when they're needed again without computing the hashes of all the keys in the segment. That's why I don't think caching is worth it.

Reply:
It's really difficult to talk about trees without diagrams ... so I've tried to explain what I mean the old fashioned way.

![IMG_20200129_095355](https://user-images.githubusercontent.com/765332/73346358-a4b6a580-427d-11ea-9955-4217aff22cb4.jpg)

Reply:
Do you keep the information that node 1 is computed from the hashes of keys 3, 4, 5 anywhere?

Reply:
Summarising a follow up zulip conversation. My proposal to cache node hashes won't work, as it does not take into account entries that are not stored in-memory (i.e persisted values). Furthermore, because all stored values need to be part of the CR, the tree should always be lazily created at the start of CR.